compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -390,45 +475,195 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
+struct TileAndFuseResult {
+  SmallVector<Operation *> tiledAndFusedOps;
+  SmallVector<scf::ForOp> loops;
+};
+
+struct TileAndFuseDispatchUsingSCFForOp
+    : public OpInterfaceRewritePattern<TilingInterface> {
+  /// Construct a generic pattern applied to all TilingInterface ops.
+  TileAndFuseDispatchUsingSCFForOp(MLIRContext *context,
+                                   linalg::LinalgTilingOptions options,
+                                   linalg::LinalgTransformationFilter filter,
+                                   PatternBenefit benefit = 1)
+      : OpInterfaceRewritePattern<TilingInterface>(context, benefit),
+        tilingPattern(context, options, filter) {}
+
+  /// Construct a generic pattern applied to `opName`.
+  TileAndFuseDispatchUsingSCFForOp(StringRef opName, MLIRContext *context,
+                                   linalg::LinalgTilingOptions options,
+                                   linalg::LinalgTransformationFilter filter,
+                                   PatternBenefit benefit = 1)
+      : OpInterfaceRewritePattern<TilingInterface>(context, benefit),
+        tilingPattern(context, options, filter) {}
+
+  FailureOr<TileAndFuseResult> returningMatchAndRewrite(
+      TilingInterface op, PatternRewriter &rewriter) const;
+
+  LogicalResult matchAndRewrite(TilingInterface op,
+                                PatternRewriter &rewriter) const override {
+    return returningMatchAndRewrite(op, rewriter);
+  }
+
+ private:
+  TileDispatchUsingSCFForOp tilingPattern;
+};
+}  // namespace
+
+/// Adds operands of `op` that are produced by `tensor.extract_slice` ops to
+/// `candidateSliceOps`.
+static void addCandidateSlices(
+    Operation *op, SmallVector<tensor::ExtractSliceOp> &candidateSliceOps) {
+  for (OpOperand &operand : op->getOpOperands()) {
+    auto sliceOp = operand.get().getDefiningOp<tensor::ExtractSliceOp>();
+    if (!sliceOp) continue;
+    candidateSliceOps.push_back(sliceOp);
+  }
+}
+
+/// Returns a list of `tensor.extract_slice` ops from `candidateSliceOps` that
+/// have the same producer. Removes those from the `candidateSliceOps`. Method
+/// is guaranteed to reduce the size of `candidateSliceOps` by at least one.
+static SmallVector<tensor::ExtractSliceOp> getSlicesFromSameProducer(
+    SmallVector<tensor::ExtractSliceOp> &candidateSliceOps) {
+  auto sliceOp = candidateSliceOps.front();
+  auto producer = sliceOp.getSource().getDefiningOp<TilingInterface>();
+  if (!producer) {
+    candidateSliceOps =
+        llvm::to_vector(llvm::makeArrayRef(candidateSliceOps).drop_front());
+    return {};
+  }
+  SmallVector<tensor::ExtractSliceOp> slicesFromSameProducer;
+  SmallVector<tensor::ExtractSliceOp> remainingCandidates;
+  slicesFromSameProducer.push_back(sliceOp);
+  for (auto otherSliceOp : llvm::makeArrayRef(candidateSliceOps).drop_front()) {
+    auto currProducer = otherSliceOp.getSource().getDefiningOp();
+    if (currProducer == producer.getOperation()) {
```

Comment:
Will there be a case where a single producer is used in multiple slice ops, but those slice ops slice the producer's result into different shapes?

Reply:
In the general case yes. But it is hard to enforce that this doesnt happen and to handle all cases appropriately. The general tile and fuse mechanism will have to assume some properties, like "all producers has compatible parallel loops as the consumer" or not, which then implies that all slices are the same. I can do some checking though. Will add that.

Reply:
I need to think about what checks I can add here that make sense that work for `TilingInterface`. For now I think we are OK just making this assumption. I'll give this some thought and send a follow up.